### PR TITLE
Prohibit indexes on non-single expressions.

### PR DESCRIPTION
**Pull request by vpetrovykh**
Ensure that the index expression cardinality is never more than ONE.

Fixes: #1320

### CODE REVIEW

File: tests/test_edgeql_ddl.py [hunk 1]
```diff
@@ -5113,3 +5113,41 @@ async def test_edgeql_ddl_required_08(self):
...
+
+    async def test_edgeql_ddl_index_01(self):
+        with self.assertRaisesRegex(
+            edgedb.ResultCardinalityMismatchError,
+            r"the index expression has cardinality MANY"
+        ):
+            await self.con.execute(r"""
+                CREATE TYPE Foo {
+                    CREATE MULTI PROPERTY a -> int64;
+                    CReATE INDEX ON (.a);
```
Comment (elprans) [inline tests/test_edgeql_ddl.py:5125]: Please fix capitalization here.

File: tests/test_edgeql_ddl.py [hunk 2]
```diff
@@ -5113,3 +5113,41 @@ async def test_edgeql_ddl_required_08(self):
...
+    async def test_edgeql_ddl_index_02(self):
+        with self.assertRaisesRegex(
+            edgedb.ResultCardinalityMismatchError,
+            r"the index expression has cardinality MANY"
+        ):
+            await self.con.execute(r"""
+                CREATE TYPE Foo {
+                    CREATE PROPERTY a -> int64;
+                    CREATE PROPERTY b -> int64;
+                    CReATE INDEX ON ({.a, .b});
```
Comment (elprans) [inline tests/test_edgeql_ddl.py:5138]: .. and here.

File: tests/test_edgeql_ddl.py [hunk 3]
```diff
@@ -5113,3 +5113,41 @@ async def test_edgeql_ddl_required_08(self):
...
+    async def test_edgeql_ddl_index_03(self):
+        with self.assertRaisesRegex(
+            edgedb.ResultCardinalityMismatchError,
+            r"the index expression has cardinality MANY"
+        ):
+            await self.con.execute(r"""
+                CREATE TYPE Foo {
+                    CREATE PROPERTY a -> int64;
+                    CREATE PROPERTY b -> int64;
+                    CReATE INDEX ON (array_unpack([.a, .b]));
```
Comment (elprans) [inline tests/test_edgeql_ddl.py:5151]: .. and here

File: edb/schema/indexes.py [hunk 4]
```diff
@@ -302,6 +302,16 @@ def compile_expr_field(
                     singletons=frozenset(singletons),
                 ),
             )
+
+            # Check that the inferred cardinality is no more than 1
+            expr_card = expr.irast.cardinality
+            if expr_card.is_multi():
+                schema_card = expr_card.to_schema_value()[1]
+                raise errors.ResultCardinalityMismatchError(
+                    f'the index expression has cardinality {schema_card}, '
```
Comment (elprans) [inline edb/schema/indexes.py:311]: Please use `possibly more than one element returned by an expression where only singletons are allowed` like we do in all other places where we require singleton expressions.

File: tests/test_edgeql_ddl.py (hunk 1 above)
Comment (vpetrovykh) [reply tests/test_edgeql_ddl.py:5125]: OK

File: tests/test_edgeql_ddl.py (hunk 2 above)
Comment (vpetrovykh) [reply tests/test_edgeql_ddl.py:5138]: OK

File: tests/test_edgeql_ddl.py (hunk 3 above)
Comment (vpetrovykh) [reply tests/test_edgeql_ddl.py:5151]: OK

File: edb/schema/indexes.py (hunk 4 above)
Comment (vpetrovykh) [reply edb/schema/indexes.py]: Updated the error message

**Comment by 1st1** [comment]
Please also update the changelog :)

**Pull request merged**